### PR TITLE
modules: restore exclude_implicits

### PR DESCRIPTION
**Pull request by haampie**
For users on systems with relatively old environment-modules, `hide_implicits:`
`true` is unusable (it errors), and `exclude_implicits` no longer exists,
meaning that there is no way to just create modules for explicits.

That's a regression, so this PR restores `exclude_implicits: true/false` as a
separate option.

Ref #40940

### CODE REVIEW

**Comment by haampie** [comment]
@xdelaruelle fyi, we're settling on this for the next Spack release

**Comment by xdelaruelle** [comment]
@haampie Another way would be to change `.modulerc` code not to use a minimal version check in header, but only executing code if minimal version is found.

I would be possible to do so by changing the string generated by `TclModulefileWriter`:

```tcl
if {[info exists ModuleTool] && $ModuleTool eq {Modules} && [versioncmp $ModuleToolVersion 4.7] >= 0} {
    module-hide --soft --hidden-loaded <module to hide>
}
```

This *minimal version test* is described [here](https://modules.readthedocs.io/en/latest/MIGRATING.html#determining-module-implementation-and-version). Such code works even on old Modules version.

**Comment by haampie** [comment]
Would that also work if lmod parses this .modulerc file? The other "issue" is that lmod used to work pretty well with tcl modules, but it doesn't understand `module-hide` config I  think.

I didn't check, but the other alternative is `.modulename` for hiding modules, is there a downside to that?

In any case, my preference for now is to split them as separate options (exclude / hide) in Spack 0.21, s.t. it's less breaking.

**Comment by xdelaruelle** [comment]
> Would that also work if lmod parses this .modulerc file? The other "issue" is that lmod used to work pretty well with tcl modules, but it doesn't understand `module-hide` config I think.

We can improve this conditional statement to provide a branch that will fit for Lmod:

```tcl
if { [info exists ::env(LMOD_VERSION_MAJOR)] } {
   hide-version <module to hide>
} elseif {[info exists ModuleTool] && $ModuleTool eq {Modules} && [versioncmp $ModuleToolVersion 4.7] >= 0} {
    module-hide --soft --hidden-loaded <module to hide>
}
```

> I didn't check, but the other alternative is `.modulename` for hiding modules, is there a downside to that?

I am not in favor of `.modulename` as it requires:
* to rename module file if it becomes explicit
* update modules depending on this module if it becomes explicit (to change "module load .name" into "module load name")

> In any case, my preference for now is to split them as separate options (exclude / hide) in Spack 0.21, s.t. it's less breaking.

This is the good approach as I think you will release shortly before SC and I do not have a lot of time now to provide new code or extensively test things.

**Comment by haampie** [comment]
Regarding renaming when implicit/explicit: I agree, but we hit a usability issue after the "hide implicits PR", namely that `autoload: direct` + `hide_implicits: true` almost always requires long hashes to be appended to guard against module name clashes. The current solution is to always have a fixed length hash suffix for implicit modules, while explicit ones are configurable length `hash_length: N`.

So in short, we are already renaming.

We can iterate on it (e.g. always generate modules with hash suffix but use aliases for explicit modules), but at the same time I don't want it to become overly complex. Let's discuss after the release this week :)

File: lib/spack/spack/schema/modules.py [hunk 1]
```diff
@@ -188,52 +188,3 @@
...
-    """Change blacklist/whitelist to exclude/include.
-
-    Arguments:
-        data (dict): data from a valid modules configuration.
-        key_translations (dict): A dictionary of keys to translate to
-            their respective values.
-
-    Return:
-        (bool) whether anything was changed in data
-    """
```
Comment (alalazo) [inline lib/spack/spack/schema/modules.py:207]: Did we ever warn about these properties? Also, if we remove, we should probably remove the attributes from the schema above too.

Comment (haampie) [reply lib/spack/spack/schema/modules.py:207]: there was just one ;p used to be black/white -> exclude/include etc which were correctly deprecated

Comment (haampie) [reply lib/spack/spack/schema/modules.py:207]: Ooh, you mean they are in the scheme. Didn't see that.

Comment (haampie) [reply lib/spack/spack/schema/modules.py:207]: @alalazo do you wanna submit a PR that removes them? 0.20 release notes say they were removed already ;p

Comment (alalazo) [reply lib/spack/spack/schema/modules.py:207]: Ok, will do it

**Review by alalazo — APPROVED** [review]
